### PR TITLE
feat(cubestore): Introduce support for DATE_ADD

### PR DESCRIPTION
**Pull request by ovr**
Hello!

Thanks

### CODE REVIEW

File: rust/cubestore/src/queryplanner/udfs.rs [hunk 1]
```diff
@@ -180,6 +189,108 @@ impl CubeScalarUDF for UnixTimestamp {
...
+                            "First argument of `DATE_PART` must be non-null scalar TimestampNanosecond without timezone"
+                                .to_string(),
+                        ));
+                    }
+                };
+
+                match &inputs[1] {
+                    ColumnarValue::Scalar(scalar) => match scalar {
+                        ScalarValue::IntervalYearMonth(Some(v)) => {
+                            let years_to_add = (*v as f64 / 12_f64).floor() as i32;
```
Comment (ilya-biryukov) [inline rust/cubestore/src/queryplanner/udfs.rs:249]: Use integer division here without conversion into float

File: rust/cubestore/src/queryplanner/udfs.rs [hunk 2]
```diff
@@ -180,6 +189,108 @@ impl CubeScalarUDF for UnixTimestamp {
...
+            fun: Arc::new(|inputs| {
+                if inputs.len() != 2 {
+                    return Err(DataFusionError::Plan(
+                        "Expected two arguments in DATE_ADD".to_string(),
+                    ));
+                }
+
+                let mut result_date = match &inputs[0] {
+                    ColumnarValue::Scalar(scalar) => match (scalar.get_datatype(), scalar) {
+                        // Right now, We support only TimeUnit::Nanosecond without TZ
```
Comment (ilya-biryukov) [inline rust/cubestore/src/queryplanner/udfs.rs:226]: Use `we` with lowercase w

File: rust/cubestore/src/queryplanner/udfs.rs [hunk 3]
```diff
@@ -180,6 +189,108 @@ impl CubeScalarUDF for UnixTimestamp {
...
+                                .to_string(),
+                        ));
+                    }
+                };
+
+                match &inputs[1] {
+                    ColumnarValue::Scalar(scalar) => match scalar {
+                        ScalarValue::IntervalYearMonth(Some(v)) => {
+                            let years_to_add = (*v as f64 / 12_f64).floor() as i32;
+                            let months_to_add = (*v - (years_to_add * 12)) as u32;
```
Comment (ilya-biryukov) [inline rust/cubestore/src/queryplanner/udfs.rs:250]: nit: replace with `*v % 12`. It's shorter.

File: rust/cubestore/src/queryplanner/udfs.rs [hunk 4]
```diff
@@ -180,6 +189,108 @@ impl CubeScalarUDF for UnixTimestamp {
...
+                            let mut month = result_date.month();
+
+                            if month + months_to_add > 12 {
+                                year += 1;
+                                month = (month + months_to_add) - 12;
+                            } else {
+                                month += months_to_add;
+                            }
+
+                            result_date = result_date.with_year(year).unwrap();
```
Comment (ilya-biryukov) [inline rust/cubestore/src/queryplanner/udfs.rs:262]: nit: add `assert!(month <= 12)` here. It's good to have extra checks in case the code changes later

File: rust/cubestore/src/queryplanner/udfs.rs [hunk 5]
```diff
@@ -180,6 +189,108 @@ impl CubeScalarUDF for UnixTimestamp {
...
+                            result_date = result_date.with_month(month).unwrap();
```
Comment (ilya-biryukov) [inline rust/cubestore/src/queryplanner/udfs.rs:263]: Don't unwrap, return errors instead.

File: rust/cubestore-sql-tests/src/tests.rs [hunk 6]
```diff
@@ -2660,6 +2661,40 @@ async fn float_index(service: Box<dyn SqlClient>) {
     }
 }
 
+async fn date_add(service: Box<dyn SqlClient>) {
+    let r = service
+        .exec_query(
+            "SELECT
+            DATE_ADD(CAST('2021-01-01T00:00:00Z' as TIMESTAMP), INTERVAL '1 second'),\
```
Comment (ilya-biryukov) [inline rust/cubestore-sql-tests/src/tests.rs:2668]: Could you add a few more tests?
At least these corner cases should be covered:
- Add one day to Jun 30
- Add one month to Dec
- Add one day to Feb 28 on a leap year and non-leap year
- add one year to Feb 29 on a leap year
- add one minute to 23:59:00
- add one month to Jan 30
- add one month to Jan 29 on a non-leap year and a leap year


Is there a way to get negative values in the intervals?
If no, please add assertions.
If yes, please add tests.

**Review by ilya-biryukov — CHANGES_REQUESTED** [review]
Thanks! Mostly LG, but let's ensure this covers corner cases

File: rust/cubestore/src/queryplanner/udfs.rs (hunk 5 above)
Comment (ovr) [reply rust/cubestore/src/queryplanner/udfs.rs:263]: `Option<T>` doesnt support map_err and it's not possible to implement impl From<NoneError> to DFErrror, it's why I wrote a small fn helper

````rust
fn datetime_safety_unwrap(opt: Option<DateTime<Utc>>) -> Result<DateTime<Utc>, DataFusionError> {
    if opt.is_some() {
        return Ok(opt.unwrap());
    }

    return Err(DataFusionError::Internal(
        "Unable to calculate operation between DateTime and Interval".to_string(),
    ));
}
````

File: rust/cubestore/src/queryplanner/udfs.rs (hunk 3 above)
Comment (ovr) [reply rust/cubestore/src/queryplanner/udfs.rs]: ✅

File: rust/cubestore/src/queryplanner/udfs.rs (hunk 2 above)
Comment (ovr) [reply rust/cubestore/src/queryplanner/udfs.rs]: ✅

File: rust/cubestore/src/queryplanner/udfs.rs (hunk 1 above)
Comment (ovr) [reply rust/cubestore/src/queryplanner/udfs.rs]: ✅
